pkg/crow/crow.go

```diff
@@ -0,0 +1,309 @@
+// Package crow implements a correctness checker tool similar to Loki Canary.
+// Inspired by Cortex test-exporter.
+package crow
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"math"
+	"math/rand"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/api"
+	promapi "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+)
+
+// Config for the Crow metrics checker.
+type Config struct {
+	PrometheusAddr  string // Base URL of Prometheus server
+	GenerateSamples int    // Number of samples to generate
+	UserID          string // User ID to use when querying.
+	ExtraSelectors  string // Extra selectors for queries, i.e., cluster="prod"
+
+	// Querying Params
+
+	QueryTimeout  time.Duration // Timeout for querying
+	QueryDuration time.Duration // Time before and after sample to search
+	QueryStep     time.Duration // Step between samples in search
+
+	// Validation Params
+
+	MaxValidations    int           // Maximum amount of times to search for a sample
+	MaxTimestampDelta time.Duration // Maximum timestamp delta to use for validating.
+	ValueEpsilon      float64       // Maximum epsilon to use for validating.
+}
+
+// RegisterFlags registers flags for the config to the given FlagSet.
+func (c *Config) RegisterFlags(f *flag.FlagSet) {
+	f.StringVar(&c.PrometheusAddr, "prometheus-addr", DefaultConfig.PrometheusAddr, "Root URL of the Prometheus API to query against")
+	f.IntVar(&c.GenerateSamples, "generate-samples", DefaultConfig.GenerateSamples, "Number of samples to generate when being scraped")
+	f.StringVar(&c.UserID, "user-id", DefaultConfig.UserID, "UserID to attach to query. Useful for querying multi-tenated Cortex.")
+	f.StringVar(&c.ExtraSelectors, "extra-selectors", DefaultConfig.ExtraSelectors, "Extra selectors to include in queries, useful for identifying different instances of this job.")
+
+	f.DurationVar(&c.QueryDuration, "query-timeout", DefaultConfig.QueryTimeout, "timeout for querying")
+	f.DurationVar(&c.QueryDuration, "query-duration", DefaultConfig.QueryDuration, "time before and after sample to search")
+	f.DurationVar(&c.QueryDuration, "query-step", DefaultConfig.QueryStep, "step between samples when searching")
```

Comment:
This and `query-timeout` have the wrong field

```suggestion
	f.DurationVar(&c.QueryStep, "query-step", DefaultConfig.QueryStep, "step between samples when searching")
```

Reply:
whoops :) thanks